### PR TITLE
lib: os: fdtable: Add syscalls for POSIX read/write/close operations

### PR DESCRIPTION
**Pull request by pfalcon**
lib: os: fdtable: Add syscalls for POSIX read/write/close operations
    
    Fdtable is a kernel-level structure, so functions accessing it, like
    read(), write(), close(), should be syscalls. This conversion however
    emphasizes a difference between "syscall" and "C standard library
    function". C stdlib functions should be represented by just a
    standard linkage symbol, and can be used even without prototype.
    Syscall is however a special code sequence, which may require special
    attributes and thus a prototype for usage. To deal with these
    distinction, we define a convention that "sys_name" is a syscall,
    while "name" is a normal C function (implemented as a wrapper around
    syscall). (And we also need to define "_name" to interface with Newlib
    (actually, just "_write".))
    
    Fixes: #25407

### CODE REVIEW

File: lib/os/fdtable.c [hunk 1]
```diff
@@ -167,27 +170,66 @@ int z_alloc_fd(void *obj, const struct fd_op_vtable *vtable)
...
 		return -1;
 	}
 
 	return fdtable[fd].vtable->write(fdtable[fd].obj, buf, sz);
 }
+
+#ifdef CONFIG_USERSPACE
+ssize_t z_vrfy_sys_write(int fd, const void *buf, size_t sz)
+{
+	Z_OOPS(Z_SYSCALL_MEMORY_READ(buf, sz));
```
Comment (jukkar) [inline lib/os/fdtable.c:216]: Don't we need to make a copy of the buf data here?

Comment (pfalcon) [reply lib/os/fdtable.c:216]: No, why?

Comment (jukkar) [reply lib/os/fdtable.c:216]: Just looking the existing calls in sockets.c, but indeed we do not copy the data buffer there either.

Comment (pfalcon) [reply lib/os/fdtable.c:216]: We need to copy only stuff which may lead to security issues. E.g., if some memory block contains pointers, we need to copy those pointers (not necessarily the entire memory block!), then validate the copies, then let the kernel use copies for dereferencing.

(Yeah, yeah, someone looking for side-channel (e.g., timing) attacks hardening may copy more stuff, e.g. control-dependent data. Good luck with that to whoever will do that.)

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
this code allows all threads on the system to use all file descriptors with these APIs.

**Comment by pfalcon** [comment]
This is split off from https://github.com/zephyrproject-rtos/zephyr/pull/25443, for read/write/close syscalls which didn't lead to controversy and need for further work. Happy to see it pass CI.

**Comment by pfalcon** [comment]
> this code allows all threads on the system to use all file descriptors with these APIs.

Yes, this is addition to the POSIX subsystem, which follows POSIX semantics. This also follows your suggestion to split off parts not related to ioctl into a separate PR.

File: lib/os/fdtable.c (hunk 1 above)
Comment (cfriedt) [reply lib/os/fdtable.c:216]: The contents of buf are really application dependent, and generally are not used by the kernel, unless it is a kernel type of socket / block / char device.

Validating the application has access to the memory should be sufficient.

File: tests/net/socket/tcp_rw/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (cfriedt) [inline tests/net/socket/tcp_rw/CMakeLists.txt:1]: Awww... why didn't you use `socketpair(2)`? ... jk. ;-)

**Review by cfriedt — APPROVED** [review]
Looks good to me. Just one question though - do you think it might be possible to get rid of `posix/unistd.h` at some point?

**Comment by cfriedt** [comment]
> this code allows all threads on the system to use all file descriptors with these APIs.

From what I understand, it is unfortunately a pre-existing condition. I agree though, it's an issue that should be addressed, but maybe in a separate ticket.

**Pull request closed without merging**